Add a pointer symbolization plugin

zardus opened the pull request:
This moves the ad-hoc pointer symbolization functionality into an angr plugin. By using `state.symbolizer`, one can ensure that every pointer written to memory turns into a symbolic expression representing its offset from a per-page symbolic variable.

A lot of effort was made to make this as fast as possible, while maintaining generality.

zardus commented:
CI is failing because some naughty people didn't add binsync to the docs.

ltfish commented on angr/project.py:
```diff
@@ -693,28 +693,16 @@ def __repr__(self):
-        if self._is_java_project is None:
-            self._is_java_project = isinstance(self.arch, ArchSoot)
-        return self._is_java_project
+    def __getattr__(self, a):
```
This is magic -- why can't we have them initialized inside `__init__()`?

zardus replied:
I don't know this code, and it maintains the semantics. I can change it to __init__, but I'm not going to chase down whatever java issues arise. My suggestion is to do this first, and do that change under a different PR.

ltfish replied:
I'm moving it to `__init__()` and as far as I know nothing should break. CI will tell us if I'm right.

ltfish commented on angr/sim_state.py:
```diff
@@ -374,20 +361,20 @@ def _set_plugin_state(self, plugin, inhibit_init=False):
-        Indicates if the project's main binary is a Java Archive.
-        """
-        return self.project and self.project.is_java_project
+    def __getattr__(self, a):
```
Again, this is too much magic...

zardus replied:
This magic was already here, and I'd rather fix it under a different PR. This way, at least it doesn't make self.arch account for 5% of our runtime (though it still takes a chunk).

ltfish replied:
I'm moving it to `__init__()` and see if that works.

ltfish commented on angr/state_plugins/symbolizer.py:
```diff
@@ -0,0 +1,252 @@
+                data,
+                base=aligned_base,
+                prefix=mo.bytes_at(mo.base, mo.length - remaining_len) if aligned_base != mo.base else b"",
+                skip=() if storage is self.state.memory else (self.state.arch.ip_offset)
```
Isn't `(self.state.arch.ip_offset)` just returning the expression itself? I assume `skip` wants a collection instead.

zardus replied:
good catch! no wonder i still had issues with IPs being symbolic :-)

ltfish commented on angr/state_plugins/symbolizer.py:
```diff
@@ -0,0 +1,252 @@
+        sc.page_symbols = dict(sc.page_symbols)
+        return sc
+
+from angr.sim_state import SimState
```
Curious: Why not using indirect import?

zardus replied:
I copied the base code from SimLogger or something. I also like the indirect import better. One sec

ltfish commented on angr/storage/paged_memory.py:
```diff
@@ -490,11 +489,19 @@ def load_objects(self, addr, num_bytes, ret_on_segv=False):
 
     def _create_page(self, page_num, permissions=None):
-        return Page(
+        if self.state:
```
Nitpick: Explicit is better than implicit -- I prefer `if self.state is not None` here.

ltfish commented:
> CI is failing because some naughty people didn't add binsync to the docs.

Oops just fixed.

rhelmot commented on angr/sim_state.py:
```diff
@@ -296,19 +296,6 @@ def addr(self):
-        return self._options
-
-    @options.setter
-    def options(self, v):
```
this looks like it was a transitional thing that we never really started adding deprecation warnings for the other things. Are we ok with breaking options setting?

ltfish replied:
I'm also not sure why we want to remove it instead of adding a deprecation warning. It was originally my fault that I did not add the warning btw.

zardus replied:
Can we have a property setter without having a property getter? `self.options` being dynamic takes an insane amount of time. I think we should just pull the bandaid on this one.

ltfish replied:
A bigger problem is why we are calling `state.options` for so many times. The same applies to `state.arch`. I profiled the former at some point and felt we could move some options out of hot paths/code regions.

Lukas-Dresel commented on angr/state_plugins/symbolizer.py:
```diff
@@ -0,0 +1,252 @@
+
+            data = mo.bytes_at(aligned_base, remaining_len, allow_concrete=True)
+            if not mo.is_bytes:
+                data = self.state.solver.eval_one(data, cast_to=bytes)
```
`data.args[0].to_bytes(data.args[1] // 8, 'big')` should do the trick

Lukas-Dresel replied:
```
In [8]: data = claripy.BVV(b'asdkl;fjalksdgjlkafsjalfk;ja')                                                                                                                                                 

In [9]: data.args[0].to_bytes(data.args[1] // 8, 'big')                                                                                                                                                     
Out[9]: b'asdkl;fjalksdgjlkafsjalfk;ja'
```

Lukas-Dresel commented on angr/state_plugins/symbolizer.py:
```diff
@@ -0,0 +1,252 @@
+                data,
+                base=aligned_base,
+                prefix=mo.bytes_at(mo.base, mo.length - remaining_len) if aligned_base != mo.base else b"",
+                skip=() if storage is self.state.memory else (self.state.arch.ip_offset,)
```
This is the only use of skip and it's for a single offset only, there should be a better way of handling this I think.

Lukas-Dresel commented on angr/state_plugins/symbolizer.py:
```diff
@@ -0,0 +1,252 @@
+        for i, p_id in enumerate(self.state.memory.mem._pages):
+            if i % 100 == 0:
+                l.info("%s/%s memory pages symbolized", i, len(self.state.memory.mem._pages))
+            addr_start = self.state.memory.mem._page_addr(p_id)
```
Look, someone did find the functions :)

zardus commented on angr/storage/paged_memory.py:
```diff
@@ -788,7 +794,7 @@ def replace_memory_object(self, old, new_content):
         """
 
-        if old.object.size() != new_content.size():
+        if (old.object.size() if not old.is_bytes else len(old.object)*self.state.arch.byte_width) != new_content.size():
```
I don't think we'd have bytes in `old` for architectures with non-8-bit bytes. I'd rather not complicate it for that rare case unless it shows up as an issue.

zardus commented on angr/state_plugins/symbolizer.py:
```diff
@@ -0,0 +1,252 @@
+        #   return
+
+        write_expr = self.state.inspect.mem_write_expr
+        byte_expr = self.state.solver.eval_one(self.state.inspect.mem_write_expr, cast_to=bytes).rjust(write_expr.length//8)
```
IIRC, that's there specifically because it was incorrectly justified out of claripy...

zardus commented on angr/engines/successors.py:
```diff
@@ -208,7 +208,10 @@ def _manage_callstack(state):
         else:
-            while state.solver.is_true(state.regs._sp > state.callstack.top.stack_ptr):
+            while True:
+                cur_sp = state.solver.max(state.regs._sp) if state.has_plugin('symbolizer') else state.regs._sp
```
Okay, I took it out to remember why I put it in, and if we have a symbolic stack pointer, that `is_true` doesn't return True, and the callstack tracking code melts down.

zardus commented:
Alright, I went through and cleaned this up a bit, parameterized PAGE_SIZE, and cleaned up some other stuff a bit. CI should pass. If there are no objections, I'll merge this in so that there's less of a conflict risk before next month.
